### PR TITLE
Bug 1731222: add clusterrole for assigning userinfo to templateinstances; add clus…

### PR DESCRIPTION
**Pull request by gabemontero**
…terrolebinding to new clusterrole to k8s garbage collector

/hold

until I can bring up a cluster with clusterbot from this PR and run the template e2e suite against it 

/assign @adambkaplan 

@bparees FYI

this change is needed to allow the k8s garbage collector to complete deletion of templateinstances, otherwise deletes never complete and the e2e's fail

the SAR check that fails without these new clusterrole/clusterrolebindings is https://github.com/openshift/openshift-apiserver/blob/master/pkg/template/apiserver/registry/templateinstance/strategy.go#L148-L158

when I added debug there during manual testing, I discovered the existence of the `system:serviceaccount:kube-system:generic-garbage-collector` SA

### CODE REVIEW

**Comment by adambkaplan** [comment]
@gabemontero any idea how long `TemplateInstance` has been broken? This sounds like something we will need to backport.

**Comment by gabemontero** [comment]
> @gabemontero any idea how long `TemplateInstance` has been broken? 

the e2e's were disabled in 4.2 so *potentially* since then ... I'll need to do a sanity check against a cluster at that level to see if the various clusterrole / clusterrolebindings are such that the same issue with GC occurs

This sounds like something we will need to backport.

without question

**Comment by gabemontero** [comment]
Have this PR up via clusterbot and running template tests now

**Comment by gabemontero** [comment]
And we have clean template e2e's against this PRs cluster from clusterbot (template service broker are the ones that are skipped):

```
started: (0/1/14) "[sig-devex][Feature:Templates] templateinstance readiness test  should report ready soon after all annotated objects are ready [Suite:openshift/conformance/parallel]"

passed: (1m8s) 2020-04-27T15:17:12 "[sig-devex][Feature:Templates] templateinstance readiness test  should report ready soon after all annotated objects are ready [Suite:openshift/conformance/parallel]"

started: (0/2/14) "[sig-devex][Feature:Templates] template-api TestTemplateTransformationFromConfig [Suite:openshift/conformance/parallel]"

passed: (18.2s) 2020-04-27T15:17:30 "[sig-devex][Feature:Templates] template-api TestTemplateTransformationFromConfig [Suite:openshift/conformance/parallel]"

started: (0/3/14) "[sig-devex][Feature:Templates] templateservicebroker security test  should pass security tests [Suite:openshift/conformance/parallel]"

skip [github.com/openshift/origin/test/extended/templates/templateservicebroker_security.go:50]: The template service broker is not installed: services "apiserver" not found

skipped: (18.1s) 2020-04-27T15:17:48 "[sig-devex][Feature:Templates] templateservicebroker security test  should pass security tests [Suite:openshift/conformance/parallel]"

started: (0/4/14) "[sig-devex][Feature:Templates] templateservicebroker end-to-end test  should pass an end-to-end test [Suite:openshift/conformance/parallel]"

skip [github.com/openshift/origin/test/extended/templates/templateservicebroker_e2e.go:57]: The template service broker is not installed: services "apiserver" not found

skipped: (18s) 2020-04-27T15:18:06 "[sig-devex][Feature:Templates] templateservicebroker end-to-end test  should pass an end-to-end test [Suite:openshift/conformance/parallel]"

started: (0/5/14) "[sig-devex][Feature:Templates] templateservicebroker bind test  should pass bind tests [Suite:openshift/conformance/parallel]"

skip [github.com/openshift/origin/test/extended/templates/templateservicebroker_bind.go:45]: The template service broker is not installed: services "apiserver" not found

skipped: (17.9s) 2020-04-27T15:18:24 "[sig-devex][Feature:Templates] templateservicebroker bind test  should pass bind tests [Suite:openshift/conformance/parallel]"

started: (0/6/14) "[sig-devex][Feature:Templates] templateinstance cross-namespace test should create and delete objects across namespaces [Suite:openshift/conformance/parallel]"

passed: (41.8s) 2020-04-27T15:19:06 "[sig-devex][Feature:Templates] templateinstance cross-namespace test should create and delete objects across namespaces [Suite:openshift/conformance/parallel]"

started: (0/7/14) "[sig-devex][Feature:Templates] templateinstance creation with invalid object reports error  should report a failure on creation [Suite:openshift/conformance/parallel]"

passed: (19.2s) 2020-04-27T15:19:25 "[sig-devex][Feature:Templates] templateinstance creation with invalid object reports error  should report a failure on creation [Suite:openshift/conformance/parallel]"

started: (0/8/14) "[sig-devex][Feature:Templates] templateinstance readiness test  should report failed soon after an annotated objects has failed [Suite:openshift/conformance/parallel]"

passed: (42.5s) 2020-04-27T15:20:08 "[sig-devex][Feature:Templates] templateinstance readiness test  should report failed soon after an annotated objects has failed [Suite:openshift/conformance/parallel]"

started: (0/9/14) "[sig-devex][Feature:Templates] templateinstance impersonation tests should pass impersonation update tests [Suite:openshift/conformance/parallel]"

passed: (30.6s) 2020-04-27T15:20:38 "[sig-devex][Feature:Templates] templateinstance impersonation tests should pass impersonation update tests [Suite:openshift/conformance/parallel]"

started: (0/10/14) "[sig-devex][Feature:Templates] template-api TestTemplate [Suite:openshift/conformance/parallel]"

passed: (18s) 2020-04-27T15:20:56 "[sig-devex][Feature:Templates] template-api TestTemplate [Suite:openshift/conformance/parallel]"

started: (0/11/14) "[sig-devex][Feature:Templates] templateinstance object kinds test should create and delete objects from varying API groups [Suite:openshift/conformance/parallel]"

passed: (20.7s) 2020-04-27T15:21:17 "[sig-devex][Feature:Templates] templateinstance object kinds test should create and delete objects from varying API groups [Suite:openshift/conformance/parallel]"

started: (0/12/14) "[sig-devex][Feature:Templates] templateinstance impersonation tests should pass impersonation creation tests [Suite:openshift/conformance/parallel]"

passed: (27.2s) 2020-04-27T15:21:44 "[sig-devex][Feature:Templates] templateinstance impersonation tests should pass impersonation creation tests [Suite:openshift/conformance/parallel]"

started: (0/13/14) "[sig-devex][Feature:Templates] templateinstance impersonation tests should pass impersonation deletion tests [Suite:openshift/conformance/parallel]"

passed: (32.6s) 2020-04-27T15:22:17 "[sig-devex][Feature:Templates] templateinstance impersonation tests should pass impersonation deletion tests [Suite:openshift/conformance/parallel]"

started: (0/14/14) "[sig-devex][Feature:Templates] templateinstance security tests  should pass security tests [Suite:openshift/conformance/parallel]"

passed: (29.9s) 2020-04-27T15:22:47 "[sig-devex][Feature:Templates] templateinstance security tests  should pass security tests [Suite:openshift/conformance/parallel]"

```

/hold cancel

@adambkaplan PTAL

**Comment by gabemontero** [comment]
I've also initiated a thread with @deads2k in forum-apiserver about all this

**Comment by gabemontero** [comment]
net with @deads2k ... ACL change alone seems fine, but the question of why https://github.com/openshift/openshift-apiserver/blob/master/pkg/template/apiserver/registry/templateinstance/strategy.go#L136 did not prevent us from doing the SAR check at all when the GC is coming through needs to be understood

/hold

**Comment by gabemontero** [comment]
looks like we have an upstream bug around GC and mutating GC only metadata short circuit checks ... see https://coreos.slack.com/archives/CB48XQ4KZ/p1588008887095400 and https://kubernetes.slack.com/archives/C0EG7JC6T/p1588010128112000 for 
tl;dr

We'll need to monitor the yet to be crafted upstream fix and do something analogous to the upstream fix

**Comment by gabemontero** [comment]
UPDATE: the upstream PR which in theory should NOT require the permission add I've done in this PR is https://github.com/kubernetes/kubernetes/pull/90534

@deads2k informed me I can just copy that change into openshift-apiserver vs. picking / vendoring in the upstream change

I'll work on that officially once that PR merges, and may try intermediate forms of it on a personal cluster to verify it removes the need for the permission update in this PR.

**Comment by gabemontero** [comment]
confirmed in a local cluster that use of the current fix in https://github.com/kubernetes/kubernetes/pull/90534 in a 4.5 cluster negates the need for this PR

@deads2k @dmage FYI

also note, confirmed that we only need the openshift-apiserver changes in 4.5 ... earlier versions don't need the change when I run the template suite against those

I'll start prepping a 4.5 openshift-apiserver change for any use of `TestIsOnlyMutatingGCFields`
